core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -125,50 +134,167 @@ private[coordinator] class TransactionMetadata(val pid: Long,
     topicPartitions ++= partitions
   }
 
-  def prepareTransitionTo(newState: TransactionState): Boolean = {
+  def prepareNoTransit(): TransactionMetadataTransition =
+    prepareTransitionTo(state, producerEpoch.toShort, txnTimeoutMs, immutable.Set.empty[TopicPartition], txnStartTimestamp, txnLastUpdateTimestamp)
```

Comment:
We only allow `Empty` -> `Empty` transition, and it is a special case for adding a txnID for the first time, the inserted entry will be Empty and its pending state also Empty. Only when the pending state is cleared after the txn log write returns it is considered "created succesfully".

As for this function itself, it is only used when `initPid` is received while the current PID has an ongoing txn, i.e. it is `Ongoing`, and later we check if it is `Ongoing` we will abort the txn first and return `CONCURRENT_TXN` to let user retry.